(iterative-methods-exercises-section)=

# Iterative methods exercises


````{exercise}
:label: iterative-methods-ex-jacobi

Using a pen and calculator, calculate the first 2 iterations of the Jacobi method for solving the system of linear equations below. Use starting values of $x_i^{(0)} = 0 $ and work to 4 decimal places.

$$ \begin{align*}
    4x_1 +x_2 -x_3 +x_4 &=14,\\
    x_1 +4x_2 -x_3 -x_4 &=10,\\
    -x_1 -x_2 +5x_3 +x_4 &=-15,\\
    x_1 -x_2 +x_3 +3x_4 &=3.
\end{align*} $$

```{dropdown} Solution

The Jacobi method for this system is

$$ \begin{align*}
    x_1^{(k+1)} &= \frac{1}{4} \left( 14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)} \right), \\
    x_2^{(k+1)} &= \frac{1}{4} \left( 10 - x_1^{(k)} + x_3^{(k)} + x_4^{(k)} \right), \\
    x_3^{(k+1)} &= \frac{1}{5} \left( -15 + x_1^{(k)} + x_2^{(k)} - x_4^{(k)} \right), \\
    x_4^{(k+1)} &= \frac{1}{3} \left( 3 - x_1^{(k)} + x_2^{(k)} - x_3^{(k)} \right).
\end{align*} $$

Using starting values of $\mathbf{x}^{(0)} = \mathbf{0}$, computing the first iteration

$$ \begin{align*}
    x_1^{(1)} &= \frac{1}{4}(14 - 0 + 0 - 0) = 3.5, \\
    x_2^{(1)} &= \frac{1}{4}(10 - 0 + 0 + 0) = 2.5, \\
    x_3^{(1)} &= \frac{1}{5}(-15 + 0 + 0 - 0) = -3, \\
    x_4^{(1)} &= \frac{1}{3}(3 - 0 + 0 - 0) = 1.
\end{align*} $$

Computing the second iteration

$$ \begin{align*}
    x_1^{(1)} &= \frac{1}{4}(14 - 2.5 + (-3) - 1) = 1.875, \\
    x_2^{(1)} &= \frac{1}{4}(10 - 3.5 + (-3) + 1) = 1.125, \\
    x_3^{(1)} &= \frac{1}{5}(-15 + 3.5 + 2.5 - 1) = -2, \\
    x_4^{(1)} &= \frac{1}{3}(3 - 3.5 + 2.5 - (-3)) = 1.6667.
\end{align*} $$

```
````


In [68]:
import numpy as np

def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = jacobi(A, b, tol=1e-4)
print(f"x = {x}")

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   2.5000 |  -3.0000 |   1.0000 |
|   2 |   1.8750 |   1.1250 |  -2.0000 |   1.6667 |
|   3 |   2.3021 |   1.9479 |  -2.7333 |   1.4167 |
|   4 |   1.9755 |   1.5953 |  -2.4333 |   1.7931 |
|   5 |   2.0446 |   1.8461 |  -2.6444 |   1.6844 |
|   6 |   1.9563 |   1.7488 |  -2.5587 |   1.8153 |
|   7 |   1.9693 |   1.8251 |  -2.6220 |   1.7838 |
|   8 |   1.9423 |   1.7981 |  -2.5979 |   1.8259 |
|   9 |   1.9445 |   1.8214 |  -2.6171 |   1.8179 |
|  10 |   1.9359 |   1.8141 |  -2.6104 |   1.8313 |
|  11 |   1.9360 |   1.8213 |  -2.6163 |   1.8295 |
|  12 |   1.9332 |   1.8193 |  -2.6144 |   1.8338 |
|  13 |   1.9331 |   1.8215 |  -2.6163 |   1.8335 |
|  14 |   1.9322 |   1.8210 |  -2.6158 |   1.8349 |
|  15 |   1.9321 |   1.8217 |  -2.6163 |   1.8349 |
|  16 |   1.9318 |   1.8216 |  -2.6162 |   1.8353 |
|  17 |   1.


````{exercise} 
:label: iterative-methods-ex-gauss-seidel

Repeat {ref}`iterative-methods-ex-jacobi` using the Gauss-Seidel method.

```{dropdown} Solution

The Gauss-Seidel method for this system is

$$ \begin{align*}
    x_1^{(k+1)} &= \frac{1}{4} \left( 14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)} \right), \\
    x_2^{(k+1)} &= \frac{1}{4} \left( 10 - x_1^{(k+1)} + x_3^{(k)} + x_4^{(k)} \right), \\
    x_3^{(k+1)} &= \frac{1}{5} \left( -15 + x_1^{(k+1)} + x_2^{(k+1)} - x_4^{(k)} \right), \\
    x_4^{(k+1)} &= \frac{1}{3} \left( 3 - x_1^{(k+1)} + x_2^{(k+1)} - x_3^{(k+1)} \right).
\end{align*} $$

Using starting values of $\mathbf{x}^{(0)} = \mathbf{0}$, computing the first iteration

$$ \begin{align*}
    x_1^{(1)} &= \frac{1}{4}(14 - 0 + 0 - 0) = 3.5, \\
    x_2^{(1)} &= \frac{1}{4}(10 - 3.5 + 0 + 0) = 1.625, \\
    x_3^{(1)} &= \frac{1}{5}(-15 + 3.5 + 1.625 - 0) = -1.975, \\
    x_4^{(1)} &= \frac{1}{3}(3 - 3.5 + 1.625 - (-1.975)) = 1.0333.
\end{align*} $$

Computing the second iteration

$$ \begin{align*}
    x_1^{(1)} &= \frac{1}{4}(14 - 1.625 + (-1.975) - 1.0333) = 1.875, \\
    x_2^{(1)} &= \frac{1}{4}(10 - 1.875 + (-1.975) + 1.0333) = 1.6792, \\
    x_3^{(1)} &= \frac{1}{5}(-15 + 1.875 + 1.6792 - 1.0333) = -2.4025, \\
    x_4^{(1)} &= \frac{1}{3}(3 - 1.875 + 1.6792 - (-2.4025)) = 1.5800.
\end{align*} $$
```
````


In [69]:
import numpy as np

def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = gauss_seidel(A, b, tol=1e-4)
print(f"x = {x}")

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   1.6250 |  -1.9750 |   1.0333 |
|   2 |   2.3417 |   1.6792 |  -2.4025 |   1.5800 |
|   3 |   2.0846 |   1.7732 |  -2.5444 |   1.7444 |
|   4 |   1.9845 |   1.8039 |  -2.5912 |   1.8035 |
|   5 |   1.9504 |   1.8155 |  -2.6075 |   1.8242 |
|   6 |   1.9382 |   1.8196 |  -2.6133 |   1.8316 |
|   7 |   1.9339 |   1.8211 |  -2.6153 |   1.8342 |
|   8 |   1.9323 |   1.8216 |  -2.6160 |   1.8351 |
|   9 |   1.9318 |   1.8218 |  -2.6163 |   1.8354 |
|  10 |   1.9316 |   1.8219 |  -2.6164 |   1.8356 |
x = [ 1.93154463  1.82190482 -2.61642052  1.83559357]



````{exercise}
:label: iterative-methods-ex-sor

Repeat {ref}`iterative-methods-ex-jacobi` using the SOR method using the optimum value for the relaxation parameter.

```{dropdown} Solution

The iteration matrix for the Jacobi method for this system is

$$ \begin{align*}
    T_J &= -D^{-1}(L + U)
    = \begin{pmatrix}
        0 & \frac14 & -\frac14 & \frac14 \\
        \frac14 & 0 & -\frac14 & -\frac14 \\
        -\frac15 & -\frac15 & 0 & \frac15 \\
        \frac13 & -\frac13 & \frac13 & 0
    \end{pmatrix}.
\end{align*} $$

The eigenvalues of $T_J$ are

$$ \begin{align*}
    \lambda_1 &= 0.6054, &
    \lambda_2 &= -0.5425, &
    \lambda_3 &= -0.2853, &
    \lambda_4 &= 0.2224,
\end{align*} $$

so $\rho(T_J) = 0.6054$ and the optimum relaxation parameter

$$ \begin{align*}
    \omega &= 1 + \left( \frac{\rho(T_J)}{1 + \sqrt{1 - \rho(T_J)^2}} \right)^2
    = 1 + \left( \frac{0.6054}{1 + \sqrt{1 - 0.6054^2}} \right)^2
    = 1.1136.
\end{align*} $$

The SOR method for this system is

$$ \begin{align*}
    x_1^{(k+1)} &= -0.1136 x_1^{(k)} + \frac{1.1136}{4} \left( 14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)} \right), \\
    x_2^{(k+1)} &= -0.1136 x_1^{(k)} + \frac{1.1136}{4} \left( 10 - x_1^{(k+1)} + x_3^{(k)} + x_4^{(k)} \right), \\
    x_3^{(k+1)} &= -0.1136 + \frac{1.1136}{5} \left( -15 + x_1^{(k+1)} + x_2^{(k+1)} - x_4^{(k)} \right), \\
    x_4^{(k+1)} &= -0.1136 + \frac{1.1136}{3} \left( 3 - x_1^{(k+1)} + x_2^{(k+1)} - x_3^{(k+1)} \right).
\end{align*} $$

Using starting values of $\mathbf{x}^{(0)} = \mathbf{0}$, computing the first iteration

$$ \begin{align*}
    x_1^{(1)} &= -0.1136(0) + \frac{1.1136}{4}(14 - 0 + 0 - 0) = 3.8977, \\
    x_2^{(1)} &= -0.1136(0) + \frac{1.1136}{4}(10 - 3.8977 + 0 + 0) = 1.6989, \\
    x_3^{(1)} &= -0.1136(0) + \frac{1.1136}{5}(-15 + 3.8977 + 1.6989 - 0) = -2.0944, \\
    x_4^{(1)} &= -0.1136(0) + \frac{1.1136}{3}(3 - 3.8977 + 1.6989 - (-2.0944)) = 1.0749.
\end{align*} $$

Computing the second iteration

$$ \begin{align*}
    x_1^{(1)} &= -0.1136(3.8977) + \frac{1.1136}{4}(14 - 1.6989 + (-2.0944) - 1.0749) = 2.0994, \\
    x_2^{(1)} &= -0.1136(1.6989) + \frac{1.1136}{4}(10 - 2.0994 + (-2.0944) + 1.0749) = 1.7227, \\
    x_3^{(1)} &= -0.1136(-2.0944) + \frac{1.1136}{5}(-15 + 2.0994 + 1.7227 - 1.0749) = -2.4910, \\
    x_4^{(1)} &= -0.1136(1.7763) + \frac{1.1136}{3}(3 - 2.0994 + 1.7227 - (-2.4910)) = 1.0749.
\end{align*} $$

```
````

In [70]:
import numpy as np

def sor(A, b, omega, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Calculate omega
D = np.diag(np.diag(A))
U = -np.tril(A, -1)
L = -np.triu(A, 1)
TJ = -np.linalg.inv(D) @ (L + U)
rho = max(abs(np.linalg.eigvals(TJ)))
omega =  1 + (rho / (1 + np.sqrt(1 - rho ** 2))) ** 2
print("TJ = \n", TJ)
print("lambda = ", np.linalg.eigvals(TJ))
print(f"omega = {omega:0.4f}\n")

# Solve system
x = sor(A, b, omega, tol=1e-4)
print(f"x = {x}")

TJ = 
 [[ 0.          0.25       -0.25        0.25      ]
 [ 0.25        0.         -0.25       -0.25      ]
 [-0.2        -0.2         0.          0.2       ]
 [ 0.33333333 -0.33333333  0.33333333  0.        ]]
lambda =  [-0.60541084+0.j  0.54252238+0.j  0.28525161+0.j -0.22236315+0.j]
omega = 1.1136

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.8977 |   1.6989 |  -2.0944 |   1.0749 |
|   2 |   2.0994 |   1.7227 |  -2.4910 |   1.7763 |
|   3 |   1.9915 |   1.8349 |  -2.6012 |   1.8193 |
|   4 |   1.9299 |   1.8206 |  -2.6152 |   1.8371 |
|   5 |   1.9320 |   1.8227 |  -2.6166 |   1.8356 |
|   6 |   1.9312 |   1.8219 |  -2.6165 |   1.8357 |
|   7 |   1.9315 |   1.8219 |  -2.6165 |   1.8356 |
x = [ 1.9314956   1.82191405 -2.61644014  1.83561973]



```````{exercise}
:label: iterative-methods-ex-convergence

Compute the spectral radii for the three methods used to compute the solution to the system from {ref}`iterative-methods-ex-jacobi`. Which of the three methods would you expect to converge to a solution the fastest?

``````{dropdown} Solution

The iteration matrices for the three methods applied to this system are

$$ \begin{align*}
  T_J &= -D^{-1}(L + U) = 
  \begin{pmatrix}
    0 & \frac14 & -\frac14 & \frac14 \\
    \frac14 & 0 & -\frac14 & -\frac14 \\
    -\frac15 & -\frac15 & 0 & \frac15 \\
    \frac13 & -\frac33 & \frac13 & 0
  \end{pmatrix}, \\
  T_{GS} &= -(L + D)^{-1}U = 
  \begin{pmatrix}
    0 & -\frac14 & \frac14 & -\frac14 \\
    0 & \frac{1}{16} & \frac{3}{16} & \frac{5}{16} \\
    0 & -\frac{3}{80} & \frac{7}{80} & -\frac{3}{16} \\
    0 & \frac{7}{60} & -\frac{1}{20} & \frac14
  \end{pmatrix}, \\
  T_{SOR} &= (D + \omega L)^{-1} ((1 - \omega)D - \omega U) =
  \begin{pmatrix}
    -0.1136 & -0.2784 &  0.2784 & -0.2784 \\   
     0.0316 & -0.0361 &  0.2009 &  0.3559 \\
    -0.0183 & -0.0700 & -0.0069 & -0.2055 \\
     0.0607 &  0.1159 & -0.0262 &  0.1981 \\
  \end{pmatrix}
\end{align*} $$

The spectral radius of the iteration matrices are

$$ \begin{align*}
  \rho(T_J) &= 0.6054, \\
  \rho(T_{GS}) &= 0.3553, \\
  \rho(T_{SOR}) &= 0.1968.
\end{align*} $$

So the SOR method should converge the fastest, followed by the Gauss-Seidel method (as expected).

`````{tab-set}
````{tab-item} Python
```python
# Define linear system
A = np.array([[ 4, 1, -1, 1 ], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3 ]])
omega = 1.1136

# Extract L, D and U matrices from A
L = np.tril(A, -1)
D = np.diag(np.diag(A))
U = np.triu(A, 1)

# Compute the iteration matrices
TJ = -np.linalg.inv(D) @ (L + U)
TGS = -np.linalg.inv(L + D) @ U
TSOR = np.linalg.inv(D + omega * L) @ ((1 - omega) * D - omega * U)

# Compute the spectral radius
rho_TJ = max(abs(np.linalg.eigvals(TJ)))
rho_TGS = max(abs(np.linalg.eigvals(TGS)))
rho_TSOR = max(abs(np.linalg.eigvals(TSOR)))

# Output results
print(f"TJ = \n", TJ)
print(f"TGS = \n", TGS)
print(f"TSOR = \n", TSOR)

print(f"\nrho_TJ = {rho_TJ:0.4f}")
print(f"rho_TGS = {rho_TGS:0.4f}")
print(f"rho_TSOR = {rho_TSOR:0.4f}")
```
````

````{tab-item} MATLAB
```matlab
% Define linear system
A = [ 4, 1, -1, 1 ;
    1, 4, -1, -1 ;
    -1, -1, 5, 1 ;
    1, -1, 1, 3 ];
omega = 1.1136;

% Extract L, D and U matrices from A
L = tril(A, -1);
D = diag(diag(A));
U = triu(A, 1);

% Compute the iteration matrices
TJ = -inv(D) * (L + U);
TGS = -inv(L + D) * U;
TSOR = inv(D + omega * L) * ((1 - omega) * D - omega * U);

% Compute the spectral radius
rho_TJ = max(abs(eig(TJ)));
rho_TGS = max(abs(eig(TGS)));
rho_TSOR = max(abs(eig(TSOR)));

% Output results
if true
    TJ
    TGS
    TSOR
    fprintf("rho_TJ = %0.4f", rho_TJ)
    fprintf("rho_TGS = %0.4f", rho_TGS)
    fprintf("rho_TSOR = %0.4f", rho_TSOR)
end
```
````
`````

``````

```````


In [82]:
# Define linear system
A = np.array([[ 4, 1, -1, 1 ], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3 ]])
omega = 1.1136

# Extract L, D and U matrices from A
L = np.tril(A, -1)
D = np.diag(np.diag(A))
U = np.triu(A, 1)

# Compute the iteration matrices
TJ = -np.linalg.inv(D) @ (L + U)
TGS = -np.linalg.inv(L + D) @ U
TSOR = np.linalg.inv(D + omega * L) @ ((1 - omega) * D - omega * U)

# Compute the spectral radius
rho_TJ = max(abs(np.linalg.eigvals(TJ)))
rho_TGS = max(abs(np.linalg.eigvals(TGS)))
rho_TSOR = max(abs(np.linalg.eigvals(TSOR)))

# Output results
print(f"TJ = \n", TJ)
print(f"TGS = \n", TGS)
print(f"TSOR = \n", TSOR)

print(f"\nrho_TJ = {rho_TJ:0.4f}")
print(f"rho_TGS = {rho_TGS:0.4f}")
print(f"rho_TSOR = {rho_TSOR:0.4f}")


TJ = 
 [[ 0.         -0.25        0.25       -0.25      ]
 [-0.25        0.          0.25        0.25      ]
 [ 0.2         0.2         0.         -0.2       ]
 [-0.33333333  0.33333333 -0.33333333  0.        ]]
TGS = 
 [[ 0.         -0.25        0.25       -0.25      ]
 [ 0.          0.0625      0.1875      0.3125    ]
 [ 0.         -0.0375      0.0875     -0.1875    ]
 [ 0.          0.11666667 -0.05        0.25      ]]
TSOR = 
 [[-0.1136     -0.2784      0.2784     -0.2784    ]
 [ 0.03162624 -0.03609344  0.20089344  0.35590656]
 [-0.0182572  -0.07004398 -0.00685177 -0.20545774]
 [ 0.06068505  0.11594452 -0.02622706  0.19812051]]

rho_TJ = 0.6054
rho_TGS = 0.3553
rho_TSOR = 0.1968



```````{exercise}
:label: iterative-methods-ex-code

Write a program to calculate the solution to the system of linear equations from {ref}`iterative-methods-ex-jacobi` using the Jacobi, Gauss-Seidel and SOR methods using a convergence tolerance of $tol=10^{-6}$. How many iterations did each of the three methods take to converge to the solution?

``````{dropdown} Solution
$x_1 = 1.931507, x_2 = 1.821918, x_3 = -2.616438, x_4 = 1.835616$

- Jacobi: 28 iterations
- Gauss-Seidel: 15 iterations
- SOR: 10 iterations

`````{tab-set}
````{tab-item} Python
```python
import numpy as np

def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if np.linalg.norm(r, np.inf) < tol:
            break

    print(f"The Jacobi method took {k} iterations to converge")
    
    return x


def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if np.linalg.norm(r, np.inf) < tol:
            break

    print(f"The Gauss-Seidel method took {k} iterations to converge")

    return x


def sor(A, b, omega, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

    print(f"The SOR method took {k} iterations to converge")

    return x


 # Define system
A = np.array([[4, 1, -1, 1], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = jacobi(A, b)
x = gauss_seidel(A, b)
x = sor(A, b, omega)

print(f"\nx1 = {x[0]:0.6f}, x2 = {x[1]:0.6f}, x3 = {x[2]:0.6f}, x4 = {x[3]:0.6f}\n")
```
````

````{tab-item} MATLAB
```matlab
function x = jacobi(A, b, tol)

n = length(b);
x = zeros(n, 1);
maxiter = 100;

for k = 0 : 100
    xold = x;
    for i = 1 : n
        sum_ = 0;
        for j = 1 : n
            if j ~= i
                sum_ = sum_ + A(i,j) * xold(j);
            end
        end
        x(i) = (b(i) - sum_) / A(i,i);
    end
    r = b - A * x;
    if norm(r, inf) < tol
        break
    end

end

fprintf("The Jacobi method took %i iterations to converge", k)

end

function x = gauss_seidel(A, b, tol)

n = length(b);
x = zeros(n, 1);
maxiter = 100;
for k = 0 : 100
    for i = 1 : n
        sum_ = 0;
        for j = 1 : n
            if j ~= i
                sum_ = sum_ + A(i,j) * x(j);
            end
        end
        x(i) = (b(i) - sum_) / A(i,i);
    end
    r = b - A * x;
    if norm(r, inf) < tol
        break
    end
end

fprintf("The Gauss-Seidel method took %i iterations to converge", k)

end

function x = sor(A, b, omega, tol)

n = length(b);
x = zeros(n, 1);
for k = 0 : 100
    for i = 1 : n
        sum_ = 0;
        for j = 1 : n
            if j ~= i
                sum_ = sum_ + A(i,j) * x(j);
            end
        end
        x(i) = (1 - omega) * x(i) + omega * (b(i) - sum_) / A(i,i);
    end
    r = b - A * x;
    if norm(r, inf) < tol
        break
    end
end

fprintf("The SOR method took %i iterations to converge", k)

end

% Define system
A = [ 4, 1, -1, 1 ;
      1, 4, -1, -1 ;
     -1, -1, 5, 1 ;
      1, -1, 1, 3 ];
b = [ 14 ; 10 ; -15 ; 3];

% Solve system
tol = 1e-6;
omega = 1.1136;
x = jacobi(A, b, tol);
x = gauss_seidel(A, b, tol);
x = sor(A, b, omega, tol)
```
````

`````
``````
```````


In [72]:
import numpy as np

def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if np.linalg.norm(r, np.inf) < tol:
            break

    print(f"The Jacobi method took {k} iterations to converge")
    
    return x


def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if np.linalg.norm(r, np.inf) < tol:
            break

    print(f"The Gauss-Seidel method took {k} iterations to converge")

    return x


def sor(A, b, omega, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

    print(f"The SOR method took {k} iterations to converge")

    return x


 # Define system
A = np.array([[4, 1, -1, 1], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = jacobi(A, b)
x = gauss_seidel(A, b)
x = sor(A, b, omega)

print(f"\nx1 = {x[0]:0.6f}, x2 = {x[1]:0.6f}, x3 = {x[2]:0.6f}, x4 = {x[3]:0.6f}\n")

The Jacobi method took 28 iterations to converge
The Gauss-Seidel method took 15 iterations to converge
The SOR method took 10 iterations to converge

x1 = 1.931507, x2 = 1.821918, x3 = -2.616438, x4 = 1.835616

